experiment/cherrypicker/server.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		return err
+	}
+	baseBranch := pr.Base.Ref
+
+	// TODO: Use a whitelist for allowed base and target branches.
+	if baseBranch != "master" || baseBranch == targetBranch {
+		resp := fmt.Sprintf("Base branch (%s) needs to be 'master' and differ from target branch (%s)", baseBranch, targetBranch)
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Only members should be able to do cherry-picks.
+	ok, err := s.GithubClient.IsMember(org, commentAuthor)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		resp := fmt.Sprintf("only %s org members may request cherry picks", org)
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Cherry-pick only merged PRs.
+	if !pr.Merged {
+		resp := "cannot cherry-pick an unmerged PR"
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Clone the repo, checkout the target branch.
+	startClone := time.Now()
+	r, err := s.GitClient.Clone(ic.Repo.FullName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			s.Logger.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.Checkout(targetBranch); err != nil {
+		return err
+	}
+	s.Logger.WithField("duration", time.Since(startClone)).Info("Cloned and checked out target branch.")
+
+	// Fetch the patch from Github
+	rsp, err := http.Get(fmt.Sprintf("https://patch-diff.githubusercontent.com/raw/%s/%s/pull/%d.patch", org, repo, num))
+	if err != nil {
+		return err
+	}
+	defer rsp.Body.Close()
+	if rsp.StatusCode < 200 || rsp.StatusCode > 299 {
+		return fmt.Errorf("response status was not 2XX: %s", rsp.Status)
+	}
+	localPath := fmt.Sprintf("/tmp/%d-%s.patch", num, uuid.NewV1().String())
+	out, err := os.Create(localPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	if _, err := io.Copy(out, rsp.Body); err != nil {
+		return err
+	}
+
+	if err := r.Config("user.name", "cherrypicker"); err != nil {
+		return err
+	}
+	if err := r.Config("user.email", "cherrypicker@localhost"); err != nil {
+		return err
+	}
+
+	// Apply the patch.
+	if err := r.Apply(localPath); err != nil {
+		resp := fmt.Sprintf("PR %d failed to apply on top of branch %q: %v", num, targetBranch, err)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Push a new branch with the applied patch in bot's fork.
+	newBranch := fmt.Sprintf("cherry-pick-%d-to-%s", num, targetBranch)
+	if err := r.CheckoutNewBranch(newBranch); err != nil {
+		return err
+	}
+	if err := r.Push(s.BotName, s.BotCredentials, repo, newBranch); err != nil {
```

Comment:
If for whatever reason we have two commands for the same PR, this needs `--force`?

Reply:
And above we need to handle the branch existing

Reply:
This repo is cloned only for this event. The cloning is from a local mirror so it's very fast (~5-10 seconds).